src/mlpack/core/boost_backport/boost_backport_serialization.hpp

```diff
@@ -24,5 +25,21 @@
   #include <boost/serialization/unordered_map.hpp>
 #endif
 
+#if BOOST_VERSION <= 105800
+  /**
+   * Boost versions 1.58.0 and earlier have a different vector serialization
+   * behaivor as compared to later versions. Notably, loading a
+   * std::vector<arma::mat> does not clear the vector before the load
+   * in v1.58 and earlier; while in the later versions, the vector is cleared
+   * before loading. This causes some tests related to serialization to fail
+   * with versions 1.58. This backport solves the issue.
+   */
+  #include "mlpack/core/boost_backport/collections_load_imp.hpp"
+  #include "mlpack/core/boost_backport/collections_save_imp.hpp"
+  #include "mlpack/core/boost_backport/vector.hpp"
```

Comment:
If a user happens to do something like this:

```
#include <boost/serialization/vector.hpp>
#include <mlpack/core.hpp>
```

and the Boost version is 1.58, then this will actually not fix the issue because `vector.hpp` is included before we have a chance to include it properly.  I think it would be useful to add a warning if the Boost version is 1.58 and `vector.hpp` is already included; that could say something a little like the Armadillo warning in `mlpack/prereqs.hpp`.  What do you think?

Reply:
Ah, you're right. I completely missed that. I don't suppose there is some way we can enforce the order of includes (as opposed to just giving a warning - which most people completely ignore anyway)? I ask because you've faced exactly the same issue with armadillo. I'll add the warning for now and look into whether there is some way to enforce the include of backported vector.hpp (and no other vector.hpp) if the detected boost version is 1.58.